docs/linters.md

```diff
@@ -0,0 +1,210 @@
+It is the workhorse of our linting system, although in some cases it
+dispatches the heavy lifting to other components such as pyflakes,
+jslint, and other home grown tools.
+
+You can find the source code [here](https://github.com/zulip/zulip/blob/master/tools/lint-all).
+
+In order for our entire lint suite to run in a timely fashion, the `lint-all`
+script performs several lint checks in parallel by forking out subprocesses.  This mechanism
+is still evolving, but you can look at the method `run_parallel` to get the
+gist of how it works.
+
+### Special options
+
+You can use the `-h` option for `lint-all` to see its usage.  One particular
+flag to take note of is the `--modified` flag, which enables you to only run
+lint checks against files that are modified in your git repo.  Most of the
+"sub-linters" respect this flag, but some will continue to process all the files.
+Generally, a good workflow is to run with `--modified` when you are iterating on
+the code, and then run without that option right before commiting new code.
+
+### Lint checks
+
+The next part of this document describes the lint checks that we apply to
+various file types.
+
+#### Generic source code checks
+
+We check almost our entire codebase for trailing whitespace.  Also, we
+disallow tab (\t) characters in all but two files.
+
+We also have custom regex-based checks that apply to specific file types.
+For relatively minor files like Markdown files and JSON fixtures, this
+is the extent of our checking.
+
+#### Python code
+
+The bulk of our Python linting gets outsourced to the "pyflakes" tool.  We
+call "pyflakes" in a fairly vanilla fashion, and then we post-process its
+output to exclude certain types of errors that Zulip is comfortable
+ignoring.  One notable class of error that Zulip currently tolerates is
+unused imports.
+
+Zulip also has custom regex-based rules that it applies to Python code.
+Look for `python_rules` in the source code for `lint-all`.  Note that we
+provide a mechanism to excude certain lines of codes from these checks.
+Often, it is simply the case that our regex approach is too crude to
+correctly exonerate certain valid constructs.  In other cases, the code
+that we exempt may be deemed not worthwhile to fix.
+
+#### JavaScript code
+
+We check our JavaScript code in a few different ways:
+- We run jslint.
+- We perform custom Zulip regex checks on the code.
+- We verify that all addClass calls, with a few exceptions, explicitly
+  contain a CSS class.
+
+The last check happens via a call to `./tools/find-add-class`.  This
+particular check is a work in progress, as we are trying to evolve a
+more rigorous system for weeding out legacy CSS styles, and the ability
+to quickly introspect our JS code for `addClass` calls is part of our
+vision.
+
+#### Puppet scripts
+
+We use Puppet as our tool to manage configuration files, using
+"puppet scripts."  To lint puppet scripts, we use the "parser validate"
```

Comment:
I think they call them `puppet manifests`; would mention our config is under `puppet/zulip`